### DISCUSSION TITLE
AVRDUDE v7.0 released

### DISCUSSION THREAD

**Discussion by dl8dtl** [avrdudes/avrdude#946] [Announcements]
I just released version 7.0 of AVRDUDE, the first release being rolled from the Github infrastructure.

While I am sure there are many more things to do, and some of those might have deserved to go into the new release, it was already way behind its original schedule (March 2022) anyway. First, there were a number of local personal things that caused it to be deferred, and then there was Stefan Rueger's excellent analysis of mistakes in `avrdude.conf` in issue #915 which I wanted to clean up before a release.

Enjoy – and be assured that there will be a next release. ;-)

**Comment by MCUdude** [comment]
Excellent!
Should we perhaps upload pre-compiled binaries for Linux and MacOS as well? It's nice to have a place where known-good binaries can be downloaded.

**Reply by dl8dtl** [reply]
I think, normally the idea is that the infrastructure of the various Linux distributions as well as Macports and Brew will pick it up soon. (I updated the FreeBSD port right yesterday.)

But we can also provide some binaries here.

**Reply by MCUdude** [reply]
Great! Arduino IDE provides Avrdude for these platforms through a JSON file. It would be great if all these was present as precompiled binaries! 🤞 

(package_index.json)
``` json
{
          "name": "avrdude",
          "version": "6.3.0-arduino18",
          {
              "size": "220677",
              "checksum": "SHA-256:2e25c9e99c255d595a1072679a88ecddfa12c223b18510760bb867039f35efa5",
              "host": "arm-linux-gnueabihf",
              "archiveFileName": "avrdude-6.3.0-arduino18-armhf-pc-linux-gnu.tar.bz2",
              "url": "http://downloads.arduino.cc/tools/avrdude-6.3.0-arduino18-armhf-pc-linux-gnu.tar.bz2"
            },
            {
              "size": "231047",
              "checksum": "SHA-256:4f88bb50d2235182ed7aa9e0a1d08e4bb956378ac9569b8e1141e37ed314fb2d",
              "host": "aarch64-linux-gnu",
              "archiveFileName": "avrdude-6.3.0-arduino18-aarch64-pc-linux-gnu.tar.bz2",
              "url": "http://downloads.arduino.cc/tools/avrdude-6.3.0-arduino18-aarch64-pc-linux-gnu.tar.bz2"
            },
            {
              "size": "280072",
              "checksum": "SHA-256:df1dfd18e2e287c47232605cd4fa41751eb70df8c300aeb7a00a3a09b524a1b8",
              "host": "x86_64-apple-darwin12",
              "archiveFileName": "avrdude-6.3.0-arduino18-x86_64-apple-darwin12.tar.bz2",
              "url": "http://downloads.arduino.cc/tools/avrdude-6.3.0-arduino18-x86_64-apple-darwin12.tar.bz2"
            },
            {
              "size": "254460",
              "checksum": "SHA-256:1ae46972b276b8a54c459f87c4ff326abdad0be2b1a293d73bf86e47765eddc3",
              "host": "x86_64-linux-gnu",
              "archiveFileName": "avrdude-6.3.0-arduino18-x86_64-pc-linux-gnu.tar.bz2",
              "url": "http://downloads.arduino.cc/tools/avrdude-6.3.0-arduino18-x86_64-pc-linux-gnu.tar.bz2"
            },
            {
              "size": "244550",
              "checksum": "SHA-256:5c8cc6c17db9300e1451fe41cd7178b0442b4490ee6fdbc0aed9811aef96c05f",
              "host": "i686-linux-gnu",
              "archiveFileName": "avrdude-6.3.0-arduino18-i686-pc-linux-gnu.tar.bz2",
              "url": "http://downloads.arduino.cc/tools/avrdude-6.3.0-arduino18-i686-pc-linux-gnu.tar.bz2"
            },
            {
              "size": "329515",
              "checksum": "SHA-256:0781f4183e91a9783c2330035520144ab76b8f75c0a9f7a25877c063bc984c4d",
              "host": "i686-mingw32",
              "archiveFileName": "avrdude-6.3.0-arduino18-i686-w64-mingw32.zip",
              "url": "http://downloads.arduino.cc/tools/avrdude-6.3.0-arduino18-i686-w64-mingw32.zip"
            }
          ]
        },
```

**Reply by dl8dtl** [reply]
> It would be great if all these was present as precompiled binaries!

Well, providing them is certainly not a big deal (although by now, we'd have to compile them manually), but as far as I can tell, they used to apply their own set of patches. Not sure whether these patches are now already obviated by our recent changes – I remember some of them merely looked like hacks rather than something that we'd like to include by default. But my memory might be wrong here.

**Reply by MCUdude** [reply]
I've looked through their patched version of Avrdude, and I've tried to incorporate as many fixes they have provided over the years as possible into this repo. AFAIK they _could_ ship Avrdude 7.0 without any modifications, and no one would complain. IIRC started providing patched versions to get bug fixes and new features quicker. For instance, the Arduino UNO Wifi rev2 (ATmega4809) has been out for almost four years, but it wasn't officially supported by Avrdude since 6.4.

The reason I want all these pre-compiled binaries is that I'm using them when providing custom Arduino cores like [MightyCore](https://github.com/MCUdude/MightyCore) and [MegaCoreX](https://github.com/MCUdude/MegaCoreX). Currently, they're using the Avrdude version Arduino IDE ships with, but now that Avrdude 7.0 is released, I'd much rather want to use 7.0 than the patched-up Arduino version. I need to host them myself (the executable needs to be in a folder named `etc`), but it's very convenient to pull them straight from Github instead of cross-compiling them myself.

**Reply by dl8dtl** [reply]
It would IMHO make sense them to setup a Github action that can produce all their artifacts, in a way that is compatible with their existing distribution format.

@mariusgreuel any opinion on that?

**Reply by mariusgreuel** [reply]
> It would IMHO make sense them to setup a Github action that can produce all their artifacts

I can do this. Currently, I moved only the MSVC windows binaries to the release, but we could add any of these:
<https://nightly.link/avrdudes/avrdude/workflows/build/main>

I remember we omitted the Un*x binaries in favor of building them locally. If we ship them, we should double check the shared library dependencies.

**Reply by MCUdude** [reply]
> It would IMHO make sense them to setup a Github action that can produce all their artifacts, in a way that is compatible with their existing distribution format.

@dl8dtl that would be awesome! Then I can link directly to the binaries bundled with the 7.0 release in this repo rather than having to host them myself. The Avrdude executable is located in a folder called `bin`, and the avrdude.conf file is located in a folder called `etc`. For reference, the tar.bz2 files shown in my previous post can be downloaded to see what the structure looks like

**Reply by mariusgreuel** [reply]
@MCUdude You are interested in the `[avrdude-linux-*](https://nightly.link/avrdudes/avrdude/workflows/build/main)` binaries, correct? Is it possible that you could check whether the binaries work for you, or in other words, do the binaries use the correct dependencies (they use all shared libries)? I remember I used the armhf build once on my Raspi, and it did not work because I was still on Buster and build in on Bullseye. So, I used a Buster docker container for the cross-builds instead. I am not sure however, if that is the proper way to do when we want to build binaries for distribution.

**Reply by dl8dtl** [reply]
> The Avrdude executable is located in a folder called `bin`, and the avrdude.conf file is located in a folder called `etc`.

There appears to be one additional `avrdude` directory on top of them.

**Reply by MCUdude** [reply]
@mariusgreuel I need binaries for all major platforms that Arduino IDE can be installed onto, but I want them to be from a release and not a nightly build. By looking at the JSON snippet I posted earlier, I need v7.x binaries for `arm-linux-gnueabihf`, `aarch64-linux-gnu`, `x86_64-apple-darwin12`, `x86_64-linux-gnu`, `i686-linux-gnu` and `i686-mingw32`.

>  Is it possible that you could check whether the binaries work for you, or in other words, do the binaries use the correct dependencies (they use all shared libraries)?

I can test on an Intel Mac, a PC, an x64 Linux computer, and on a really old Rpi model B running the latest Raspbian. But shouldn't a binary "just work" when it's executed on a computer with the correct operating system? I thought libraries were statically linked and didn't need to be installed by the user separately?

**Reply by umbynos** [reply]
Hello guys, @arduino maintainer here,
Thanks to the effort you guys put in porting the avrdude project to GitHub, we found some bandwidth to look into the git history of our fork.
As some of you may know, we distribute tools and packages using a `package_index.json` file that has download links inside it, just like @MCUdude [pointed out above](https://github.com/avrdudes/avrdude/discussions/946#discussioncomment-2708358). We are planning to provide AVRDUDE 7.x in our packages using that index.

I started working on a [temporary fork](https://github.com/umbynos/avrdude/tree/dev) that will be transferred under Arduino org (but it still needs a lot of love and work). Along with @facchinm, I found the patches that we still need, starting from your main branch. They are currently hosted in my [`dev`](https://github.com/umbynos/avrdude/tree/dev) branch.

I also started to work on a [build & release CI](https://github.com/umbynos/avrdude/blob/dev/.github/workflows/release.yml) based on GH actions that is able to produce binaries as static as possible. We need static binaries because they would work on almost any machine the user is using. To do so, we are trying to cross compile using a [docker container I made](https://github.com/arduino/rp2040tools/tree/master/docker). We are using that container also to build another tools called [`rp2040tools`](https://github.com/arduino/rp2040tools) (If the container gains traction, I plan to place it to a dedicated repo in the near future).
The binaries produced can be downloaded for testing purposes from [here](https://github.com/umbynos/avrdude/actions/runs/2282132157).

My plan would be to:

- Finish the CI (currently the only platform giving me problems is macOS).
- See together with you which patches could be upstreamed.
- Move the release CI to a dedicated GH repo under Arduino organization: Ideally this repo should contain only the patches that could not be upstreamed, and the release workflow. The source code will be only on the [official avrdude repo](https://github.com/avrdudes/avrdude).


These are the lib dependencies of linux_amd64 binary, [produced by the CI](https://github.com/umbynos/avrdude/actions/runs/2282132157) I'm working on :
```
$ ldd avrdude
	linux-vdso.so.1 (0x00007ffec4dc0000)
	libm.so.6 => /lib/x86_64-linux-gnu/libm.so.6 (0x00007f4efd4db000)
	libpthread.so.0 => /lib/x86_64-linux-gnu/libpthread.so.0 (0x00007f4efd2bc000)
	libc.so.6 => /lib/x86_64-linux-gnu/libc.so.6 (0x00007f4efcecb000)
	/lib64/ld-linux-x86-64.so.2 (0x00007f4efd879000)

$ file avrdude
avrdude: ELF 64-bit LSB executable, x86-64, version 1 (SYSV), dynamically linked, interpreter /lib64/ld-linux-x86-64.so.2, for GNU/Linux 4.4.94, with debug_info, not stripped
```

The binary [downloaded from your CI](https://github.com/avrdudes/avrdude/actions/runs/2289555695) has the following deps:
```
$ ldd avrdude
./avrdude: /lib/x86_64-linux-gnu/libm.so.6: version `GLIBC_2.29' not found (required by ./avrdude)
	linux-vdso.so.1 (0x00007ffc6f931000)
	libm.so.6 => /lib/x86_64-linux-gnu/libm.so.6 (0x00007f12dbe87000)
	libelf.so.1 => /usr/lib/x86_64-linux-gnu/libelf.so.1 (0x00007f12dbc6d000)
	libusb-0.1.so.4 => not found
	libusb-1.0.so.0 => /lib/x86_64-linux-gnu/libusb-1.0.so.0 (0x00007f12dba55000)
	libhidapi-libusb.so.0 => not found
	libftdi1.so.2 => not found
	libreadline.so.8 => not found
	libc.so.6 => /lib/x86_64-linux-gnu/libc.so.6 (0x00007f12db664000)
	/lib64/ld-linux-x86-64.so.2 (0x00007f12dc225000)
	libz.so.1 => /lib/x86_64-linux-gnu/libz.so.1 (0x00007f12db447000)
	libudev.so.1 => /lib/x86_64-linux-gnu/libudev.so.1 (0x00007f12db229000)
	libpthread.so.0 => /lib/x86_64-linux-gnu/libpthread.so.0 (0x00007f12db00a000)
	librt.so.1 => /lib/x86_64-linux-gnu/librt.so.1 (0x00007f12dae02000)

$ file avrdude
avrdude: ELF 64-bit LSB shared object, x86-64, version 1 (SYSV), dynamically linked, interpreter /lib64/ld-linux-x86-64.so.2, BuildID[sha1]=76ca1e486c10d85b049edb87cf72bd60a44b709c, for GNU/Linux 3.2.0, with debug_info, not stripped
```
PS I'm on Ubuntu 18.04 LTS

**Reply by umbynos** [reply]
Quick update:

> - Finish the CI (currently the only platform giving me problems is macOS). 

Done :heavy_check_mark: 


> - See together with you which patches could be upstreamed. 

:construction: PR here:
- #981
- #982 
- #983 
- #984 

> - Move the release CI to a dedicated GH repo under Arduino organization: Ideally this repo should contain only the patches that could not be upstreamed, and the release workflow. The source code will be only on the [official avrdude repo](https://github.com/avrdudes/avrdude). 

:heavy_check_mark: ([here](https://github.com/arduino/avrdude-packing) it is, along with a test release [7.0-arduino.1](https://github.com/arduino/avrdude-packing/releases/tag/7.0-arduino.1))

**Comment by mariusgreuel** [comment]
One note for the release process: I pulled the Windows 7.0 release, and it says:
`avrdude version 6.99, URL: <https://github.com/avrdudes/avrdude>`

Before pushing the `v7.0` tag, you need to explicitly set the version number in CMakeLists.txt
<https://github.com/avrdudes/avrdude/blob/main/CMakeLists.txt#L24>

I guess it could not hurt to write up a wiki release procedure. I suppose that would help also if some other person than Jörg would like to create a release :-).

**Reply by mariusgreuel** [reply]
One more note regarding the `deploy.yml` action. If you want to make a correction after a release, it is possible to force-push the release tag. If so, however, you should delete the Github release completly, as the release and the release artifacts will always be created, and the script will choke if the release already exists.

**Reply by dl8dtl** [reply]
> One note for the release process: I pulled the Windows 7.0 release, and it says: `avrdude version 6.99, URL: <https://github.com/avrdudes/avrdude>`

OK, I fixed that, and re-tagged the release. (Deleted it completely this time.)

> I guess it could not hurt to write up a wiki release procedure.

Sure! I even missed to update the NEWS file myself this time.
